### ISSUE TITLE
General help 

### ISSUE THREAD

**Issue by AlmudenaRobledo** [neuropoly/axondeepseg#282]
We are trying to use AxonDeepSeg to segment SEM images of human hippocampus samples. I am trying to find the best way to go about this but I am running into some issues. 

The default model has a very high error on our images and we wanted to try to train a new model with our samples and myelin masks that I have made manually. These were made using the GIMP program as suggested. In order to obtain the axon segmentation from the myelin mask I have tried using the notebook "generate_axons_from_myelin" for this but I get the error:

ValueError: operands could not be broadcast together with shapes (919,1072) (1594,946,4). 

I also tried to use the fsleyes GUI to do this by uploading the manual mask but it marks what I marked as myelin as axons (blue) and the edges of the myelin sheaths as myelin (red) and ultimately doesn’t work.

I ventured to go ahead and use the manual masks I made to build the dataset to train a new model but I also ran into errors: 

ValueError: Max value == min value, ambiguous given dtype

I have to confess, as you can probably tell I have little to no computer science experience and am trying to learn as I go along. Some informed guidance would be a great help! 

Thanks!

**Comment by mathieuboudreau** [comment]
Hi @AlmudenaRobledo !

> ValueError: Max value == min value, ambiguous given dtype

We've encountered that error before (https://github.com/neuropoly/axondeepseg/issues/130) and I know I've encountered myself a few times; if I recall, I think it might occur when the segmentation results in a mask with all pixels having the same value (i.e. all white, all black). This could be happening for several reasons, but we'll help you debug it!

Would you mind sharing a sample raw image, sample mask, and the notebook(s) that caused you issues. That'll give me more information that'll be helpful before we start tackling training your dataset. Some screenshots of your problems would also be helpful (i.e. what you described that you experiences in FSLeyes). Lastly, if you could detail which version of axondeepseg you're using (i.e. how did you get axondeepseg on your computer - download button or `git clone`? If `git clone`, which git branch & git commit # are you currently using (if you know)).

Thanks!

**Comment by AlmudenaRobledo** [comment]
@mathieuboudreau Thanks for your help! 
Sample Images: 
![image](https://user-images.githubusercontent.com/52037703/74225640-65477a80-4cbb-11ea-916b-ceaefb95f2c8.png)
![mask](https://user-images.githubusercontent.com/52037703/74226695-4ba73280-4cbd-11ea-839b-6ba36cdc2e51.png)



Notebooks: 
http://localhost:8888/notebooks/guide_dataset_building_SB.ipynb

http://localhost:8889/notebooks/generate_axons_from_myelin_SB.ipynb

I downloaded Axondeepseg using git clone as described on your download instructions to the T. I'm not sure which git branch or git commit I'm using but if there is a way to find out I'd be happy to tell you.

**Comment by AlmudenaRobledo** [comment]
Lastly, the screenshots for fsleyes: 

I tried to use the fill axons function. I loaded the raw image and the sample myelin mask (with myelin in the name) and clicked the fill axons botton. These are the results I get: 
![Screen Shot 2020-02-11 at 11 08 58 AM](https://user-images.githubusercontent.com/52037703/74227898-b2c5e680-4cbf-11ea-947e-a0a914c800d0.png)
![Screen Shot 2020-02-11 at 11 13 02 AM](https://user-images.githubusercontent.com/52037703/74227900-b5284080-4cbf-11ea-9a12-e358ac8659a8.png)

**Comment by mathieuboudreau** [comment]
Great - thanks @AlmudenaRobledo ! @jcohenadad @Stoyan-I-A it sounds like maybe their error occurs because they don't have axon masks, since we've seen that error before when some masks had all the same value. Has there been more work done on the "Fill axons" feature, or is that still a work in progress?

@AlmudenaRobledo in the next few days, a feature will be merged into the master/main branch that will show a more complete example of training data, including datasets we've formated and uploaded to OSF.io. I'll tag you when that's complete, and give you brief instructions on how to test it. If that new set of updates to the notebooks work for you, then. we can start comparing our sample dataset with yours to contrast the differences.

**Comment by AlmudenaRobledo** [comment]
@mathieuboudreau  I created some axon masks using fiji imageJ and was able to build the data set without a problem. I proceeded to try out training the model but ran into some more errors:

ValueError: cannot reshape array of size 65536 into shape (512,512,1)

in the following notebook: http://localhost:8889/notebooks/training_guideline_SB.ipynb

I'll keep an eye out for the updates in case they help us move forward. Thanks again for the help.

**Comment by mathieuboudreau** [comment]
Hi @AlmudenaRobledo 

What that error is saying is that it created a 1D array of your image, and that array has 65536 values (i.e. pixels).

Then, the training tried to reconstruct an image of dimensions 512x512, but it couldn't do so.

Dividing 65536 by 512 gives 128. So that means that your original images (i.e. patches) are probably of size 256x256, not 512x512.

If you look at the "trainingset_patchsize" parameter in the training configuration of the notebook, you'll see what patch size value (A x A) it is trying to resample images to (shown below is the default values):

<img width="903" alt="Capture d’écran 2020-02-11 à 13 23 47" src="https://user-images.githubusercontent.com/1421029/74259104-eb6ebb80-4cd1-11ea-9619-c7dbf0670dd2.png">


You can see that our default is currently 512. I suspect if you change that value to 256 and rerun all cells, the error you received might be resolved. You should also check your training images to ensure that they are in fact 256x256 pixels.

Best,

Mathieu

**Comment by Stoyan-I-A** [comment]
> it sounds like maybe their error occurs because they don't have axon masks, since we've seen that error before when some masks had all the same value. Has there been more work done on the "Fill axons" feature, or is that still a work in progress?

@mathieuboudreau From what I'm see in the FSLeyes screenshots, it seems like there is an axon mask (it's just that it is where there should be myelin instead). In the current version of the plugin, I would've thought that the _fill axon_ button would fill up the entire image with an axon mask since it applies a flood-fill at the center of mass of each axon object. Strange... I'll look into this soon.

**Comment by Stoyan-I-A** [comment]
> I would've thought that the _fill axon_ button would fill up the entire image with an axon mask since it applies a flood-fill at the center of mass of each axon object. Strange... I'll look into this soon.

Well, I tested it and indeed, the generated corrected image is filled up with with an axon mask. So this algorithm works as I expected. It's just that FSLeyes will not display an image with a single intensity unless you click on the button next to the colour mapping bar.

That being said, there is an open issue about improving this feature (https://github.com/neuropoly/axondeepseg/issues/268)

**Comment by mathieuboudreau** [comment]
Hi @AlmudenaRobledo,

Have you explored a bit more your training since I mentionned the issue about your patch size? If you'd like to discuss this software more over a video call sometime, you can email me to schedule one. My email is on my profile: https://github.com/mathieuboudreau
